### ISSUE TITLE
Handle Attribute Groups better

### ISSUE THREAD

**Issue by jsbean** [dn-m/MusicXML#31]
Currently, attribute groups (e.g., `print-style-align`) are implemented as structs (e.g., `PrintStyleAlign`).

As such, they are nested into their parent type. However, the properties of attribute groups exist on the top-level of the parent type. This adds considerable complexity to the encoding/decoding of types.

Instead of nesting attribute groups as properties in complex types, we should declare attribute groups as protocols, and manually implement the properties in each type.

In other words, instead of:

```Swift
struct SomeType {
    let printStyleAlign: PrintStyleAlign
}
```

Try this:

```Swift
protocol PrintStyleAlign {
    var defaultX: Double? { get }
    var defaultY: Double? { get }
    ...
}
struct SomeType: PrintStyleAlign {
    var defaultX: Double?
    var defaultY: Double?
    ...
    var somethingElse: String?
}
```

This will require some additional keystrokes, but will make the encode/decode process much more straightforward. We can probably find/replace the current `let printStyleAlign: PrintStyleAlign` with the list of required properties.

Alternatives: Use a class-based inheritance model. I would rather not dip into reference semantics just for convenience.

**Comment by jsbean** [comment]
A more robust example is:

```Swift
protocol Position {
    var defaultX: Tenths? { get }
    var defaultY: Tenths? { get }
    var relativeX: Tenths? { get }
    var relativeY: Tenths? { get }
}
protocol PrintStyle: Position {
    var font: Font? { get }
    var color: Color? { get }
}

protocol PrintStyleAlign: PrintStyle {
    var hAlign: LeftCenterRight? { get }
    var vAlign: VAlign? { get }
}

struct Dynamics {
    ...
    // MARK: - Position
    var defaultX: Tenths?
    var defaultY: Tenths?
    var relativeX: Tenths?
    var relativeY: Tenths?
    // MARK: - PrintStyle
    var font: Font?
    var color: Color?
    // MARK: - PrintStyleAlign
    var hAlign: LeftCenterRight?
    var vAlign: VAlign?
}
```

**Comment by jsbean** [comment]
Because they are known to be attributes, it would be nice if we could hook this into the `DynamicNodeDecoding`/`Encoding`, though I don't know how that would be possible.

**Comment by jsbean** [comment]
If we were to take the protocol route, this would explode the number of properties visible at the top level of the type.

If that becomes a bit cumbersome, we could shove all attributes inside a nested `Attributes` type. We could also nest elements in their own `Elements` type.

For types which are `choice` elements at their top-level (which usually have a nested `Kind` enum), it kind of clarifies this.

```Swift
struct Key {
    struct Attributes: PrintStyle {
        // MARK: - Position
        var defaultX: Tenths?
        var defaultY: Tenths?
        var relativeX: Tenths?
        var relativeY: Tenths?
        // MARK: - PrintStyle
        var font: Font?
        var color: Color?
        var printObject: Bool?
        var number: Int?
    }
    struct Elements {
        var keyOctave: [KeyOctave]?
    }
    enum Kind {
        case traditional(Traditional)
        case nonTraditional(NonTraditional)
    }
    let attributes: Attributes
    let elements: Elements
    let kind: Kind
}
```

For ease of use, multiple initializers can be implemented to cut down on the `Kind` boilerplate:

```Swift
extension Key {
    // traditional
    init(
        fifths: Int,
        mode: Mode?,
        cancel: Cancel?, 
        attributes: Attributes, 
        elements: Elements
    ) { ... }
    // non-traditional
    init(
        step: Step,
        alter: Double,
        accidental: AccidentalValue,
        attributes: Attributes,
        elements: Elements
    ) { ... }
}
```

I haven't tried this out yet, but this may simplify the `DynamicNodeEncoding` and `DynamicNodeDecoding` configuration.

```Swift
extension Key.Attributes {
    enum CodingKeys: String, CaseIterable, CodingKey {
        case defaultX
        ...
    }
}

extension Key: DynamicNodeDecoding {
    static func nodeDecoding(for key: CodingKey) -> XMLDecoder.NodeDecoding {
        if Key.Attributes.CodingKeys.allKeys.contains(key) {
            return .attribute
        }
        return .element
    }
}
```

@bwetherfield, cast your vote:

- [ ] Keep attribute groups as structs nested in types
- [ ] Use protocols for attribute groups, explode out attributes at top level of elements
- [x] Use protocols for attributes groups, nest in an Attributes struct (furthermore, nest elements in an Elements struct)

> Something to get one's head around is the MusicXML [`Attributes`](https://usermanuals.musicxml.com/MusicXML/Content/EL-MusicXML-attributes.htm) … _element_.

**Comment by bwetherfield** [comment]
I _think_ the last way makes the most sense to me!

**Comment by bwetherfield** [comment]
Seems to have a good few benefits to it :)

**Comment by bwetherfield** [comment]
This is more specific to the case here, but did you consider having the whole of `Key` be an `enum`? So you have `Key.traditional`, `Key.nonTraditional` etc.

**Comment by jsbean** [comment]
The only issue with that is that it is beyond the capabilities of an enum (which can't have stored properties).

This ends up being a relatively common pattern (i.e., structs with a `Kind`). I would love if there was a way to automate this, but it is currently beyond my mental capacity.

**Comment by jsbean** [comment]
I spent some time trying to make the `Attributes` design work, and I haven't found a way to achieve it fully yet.

I haven't spent too much time trying to implement / correct the attribute group properties. They tend to deal with graphical aspects, which haven't come up so much yet in the LilyPond tests that I have engaged with.

That is, we can poke at it incrementally, and find ourselves in the midst of a good design (I _hope_).
